src/compiler/moduleNameResolver.ts

```diff
@@ -1868,14 +1873,24 @@ function loadModuleFromFileNoImplicitExtensions(extensions: Extensions, candidat
     }
 }
 
-function loadJSOrExactTSFileName(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
+/**
+ * This function is it’s only ever called with paths written in package.json files - never
+ * module specifiers written in source files - and so consequently it always allows the
+ * candidate to end with a TS extension (but will also try substituting a JS extension for a TS extension).
+ */
+function loadFileNameFromPackageJsonField(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {

```

Comment:
the two preceding functions start with `loadModuleFrom*`, but `candidate` isn't really a module specifier, so I don't know. I think the current suggestion is the best, but maybe `loadModuleFromPackageJsonField` isn't *too* inaccurate either. And it's parallel to the earlier functions.

(I'd probably stick with what you have, though, unless you like my suggestion a *lot* better.)

Reply:
It is definitely more filenamey than not.